File: src/Master/Modulos/AudioMod.cpp
#include "AudioMod.h"

AudioMod::AudioMod(int ninputs, int noutputs, int nInLinks, int nOutLinks):
    ninputs(ninputs),
    noutputs(noutputs),
    nInLinks(nInLinks),
    nOutLinks(nOutLinks)
{
    pthread_mutex_init(&mtx,NULL);
    pthread_mutex_lock(&mtx);

    input_lock.assign(ninputs,NULL);
    outputs.resize(noutputs*2);
    inputs.resize(ninputs*2);
    linkReady.assign(nOutLinks,1);

    pthread_mutex_unlock(&mtx);
}

int AudioMod::getOutput(ofSoundBuffer& output)
{
    if (!isOutReady())
        return -1;
    pthread_mutex_lock(&mtx);
    output=outputs;
    pthread_mutex_unlock(&mtx);
    return true;
}

int AudioMod::getOutput(ofSoundBuffer &output, int noutput)
{
    if (noutput>=noutputs || !isOutReady())
        return -1;
    pthread_mutex_lock(&mtx);
    ofSoundBuffer left, right;
    outputs.getChannel(left, 2*noutput);
    outputs.getChannel(right,2*noutput+1);
    //output.allocate(this->outputs.size(),2);
    output.setNumChannels(2);
    output.setChannel(left,0);
    output.setChannel(right,1);
    pthread_mutex_unlock(&mtx);
    return true;
}

bool AudioMod::isOutReady()
{
    pthread_mutex_lock(&mtx);
    for (int n=0;n<ninputs;n++)
        if(input_lock[n]==0)
        {
            pthread_mutex_unlock(&mtx);
            return 0;
        }
    pthread_mutex_unlock(&mtx);
    return 1;
}

bool AudioMod::noMtxOutReady()
{
    for (int n=0;n<ninputs;n++)
        if(input_lock[n]==0)
        {
            pthread_mutex_unlock(&mtx);
            return 0;
        }
    return 1;
}

/*int Modulo::getNext(std::vector<int[2]> &next)
{
    for (int n=0;n<this->next.size();n++)
    {
        next[n][0]=this->next[n][0];
        next[n][1]=this->next[n][1];
    }
    if (!isOutReady())
        return 0;
    return 1;
}
*/
int AudioMod::processInput (ofSoundBuffer inputs)
{
    if (inputs.getNumChannels()!=2*ninputs)
        return -1;
    pthread_mutex_lock(&mtx);
    input_lock.assign(ninputs,true);
    process();
    pthread_mutex_unlock(&mtx);
    return true;

}
int AudioMod::processInput(int ninput, ofSoundBuffer input)
{
    if(ninput>=ninputs || input.getNumChannels()!=2 ||input_lock[ninput]==1)
        return -1;
   pthread_mutex_lock(&mtx);
   input_lock[ninput]=1;
   ofSoundBuffer left, right;
   input.getChannel(left,0);
   input.getChannel(right,1);
   inputs.setChannel(left, ninput*2);
   inputs.setChannel(right, ninput*2+1);
   if(noMtxOutReady())
   {
       process();
       pthread_mutex_unlock(&mtx);
       return true;
   }
   else
   {
       pthread_mutex_unlock(&mtx);
       return 0;
   }
}


void AudioMod::resetBuff(int numSamples)
{
    pthread_mutex_lock(&mtx);
    outputs.allocate(numSamples, 2*noutputs);
    outputs.set(0);
    inputs.allocate(numSamples,2*ninputs);

    inputs.set(0);
    input_lock.assign(ninputs,NULL);
    pthread_mutex_unlock(&mtx);
}



int AudioMod::updateLink (int nlink, float val)
{
    if (nlink>=nInLinks)
        return -1;
    pthread_mutex_lock(&mtx);
    in_link(nlink,val);
    pthread_mutex_unlock(&mtx);
    return true;
}

float AudioMod::getLink (int nlink)
{
    if (nlink >= nOutLinks)
        return -1;
    pthread_mutex_lock(&mtx);
    float val= out_link(nlink);
    linkReady[nlink]=0;
    pthread_mutex_unlock(&mtx);
    return val;
}

bool AudioMod::isLinkReady(int nlink)
{
    if (nlink>=nOutLinks)
        return false;
    pthread_mutex_lock(&mtx);
    int val= linkReady[nlink];
    pthread_mutex_unlock(&mtx);
    return val;
}

File: src_python/Master/Audio/AudioMod/AudioMod.py
from threading import  Lock, Event
import numpy as np
from .BaseMod import BaseMod

class AudioMod (BaseMod) :
    
    def __init__ (self, ninputs : int , noutputs : int, nInlinks : int, nOutLinks:int):
         
        BaseMod.__init__ (self,nInlinks, nOutLinks)
        self.ninputs=ninputs
        self.noutputs=noutputs
        self.inputLock=np.zeros([ninputs],bool)
        self.outputBuff=np.empty([0,2*self.noutputs])
        self.inputBuff=np.empty([0,2*self.noutputs])

    #Procesa todas las entradas guardadas en el buffer que tendra que tener el tamaño adecuado
    #-1 error
    #0 procesado correctamente pero módulo no listo para output
    #>0 procesado y listo para output 
    def processInputs (self, inputs : np, e=0) : 
        self.mtx.acquire()
        if (inputs.shape[1]==2*self.ninputs):   
            self.inputBuff=inputs
            self.inputLock[:]=1
            self.process()
            self.mtx.release()
            if e!= 0 :
                e.set()
            return 1
        self.mtx.release()
        if e!= 0 :
            e.set()
        return -1 
    def processInput (self, input : np, ninput:int, e=0) :
        self.mtx.acquire()
        if (ninput >= self.ninputs or input.shape[1] != 2 or input.shape[0] != self.inputBuff.shape[0] or self.inputLock[ninput] == 1):
            self.mtx.release()
            if e!= 0 :
                e.set()
            return -1
        self.inputLock[ninput]=1
        self.inputBuff[:,2*ninput:2*ninput+2]=input[:,:]
        for x in self.inputLock :
            if x == 0:
                self.mtx.release()
                if e!= 0 :
                    e.set()
                return 0
        self.process()
        if e!= 0 :
            e.set()
        self.mtx.release()
        return 1  
    def getOutputs (self):
        self.mtx.acquire()
        outputs= np.zeros(self.outputBuff.shape)
        outputs [:]=self.outputBuff
        self.mtx.release()
        return outputs
    def getOutput (self, noutput : int):
        output=np
        self.mtx.acquire()
        output.empty ([self.outputBuff.shape[0], 2])
        output = self.outputBuff[:,2*noutput:2*noutput+2]
        self.mtx.release()
        return output
    def resetBuff (self, numSamples : int) :
        self.mtx.acquire()
        self.outputBuff.resize(numSamples, 2*self.noutputs)
        self.inputBuff.resize(numSamples, 2*self.ninputs)
        self.outputBuff[:,:]=0
        self.inputBuff[:,:]=0
        self.inputLock[:]=0
        self.mtx.release()
    def process (self):
        if self.ninputs == self.noutputs :
            self.outputBuff [:,:] =self.inputBuff[:,:]
    def isOutReady (self):
        self.mtx.acquire()
        for x in self.inputLock :
            if x == 0:
                self.mtx.release()
                return False
        self.mtx.release()
        return True
